tensorflow/core/kernels/training_ops.cc

```diff
@@ -2672,146 +2808,23 @@ class SparseApplyFtrlOp : public OpKernel {
-          }
-        }
-#undef COMPUTE_FTRL
-      } else {
-        T lr_scalar = lr.scalar<T>()();
-        T l1_scalar = l1.scalar<T>()();
-        T l2_scalar = l2.scalar<T>()();
-        T lr_power_scalar = lr_power.scalar<T>()();
-        T l2_shrinkage_scalar;
-        if (has_l2_shrinkage) {
-          l2_shrinkage_scalar = l2_shrinkage->scalar<T>()();
-        }
-
-        auto indices_vec = indices.vec<Tindex>();
-        auto var_flat = var.flat<T>();
-        auto accum_flat = accum.flat<T>();
-        auto linear_flat = linear.flat<T>();
-        auto grad_flat = grad.flat<T>();
-        const Tindex first_dim_size = accum_flat.size();
-
-        for (Tindex i = 0; i < N; i++) {
-          const Tindex index = internal::SubtleMustCopy(indices_vec(i));
-          OP_REQUIRES(ctx, FastBoundsCheck(index, first_dim_size),
-                      errors::InvalidArgument(
-                          strings::StrCat("Index ", index, " at offset ", i,
-                                          " in indices is out of range")));
-          T& a = accum_flat(index);
-          T& l = linear_flat(index);
-          T& v = var_flat(index);
-          T g;
-          if (has_l2_shrinkage) {
-            g = grad_flat(i) +
-                (static_cast<T>(2) * l2_shrinkage_scalar * var_flat(index));
-          } else {
-            g = grad_flat(i);
-          }
-
-          T updated_a = a + grad_flat(i) * grad_flat(i);
-          using Eigen::numext::pow;
-          T sigma = pow(updated_a, -lr_power_scalar) - pow(a, -lr_power_scalar);
-          if (!multiply_linear_by_lr_) {
-            sigma /= lr_scalar;
-          }
-          T updated_l = (multiply_linear_by_lr_ ? l + g * lr_scalar - sigma * v
-                                                : l + g - sigma * v);
-          v = FtrlCompute(updated_a, updated_l, lr_scalar, l1_scalar, l2_scalar,
-                          lr_power_scalar, multiply_linear_by_lr_);
-          a = updated_a;
-          l = updated_l;
-        }
-      }
-    }
+    const Device& device = ctx->template eigen_device<Device>();
+    auto indices_vec = indices.vec<Tindex>();
+    const Tindex bad_i =
+        functor::SparseApplyFtrl<Device, T, Tindex, has_l2_shrinkage>()(
+            device, var.flat_outer_dims<T>(), accum.flat_outer_dims<T>(),
+            linear.flat_outer_dims<T>(), lr.scalar<T>(), l1.scalar<T>(),
+            l2.scalar<T>(),
+            // Note: Passing l2 as a placeholder when not has_l2_shrinkage (it
+            // will not be used).
+            has_l2_shrinkage ? l2_shrinkage->scalar<T>() : l2.scalar<T>(),
+            lr_power.scalar<T>(), grad.flat_outer_dims<T>(), indices_vec,
+            inner_dim, multiply_linear_by_lr_);
+    OP_REQUIRES(
+        ctx, bad_i < 0,
+        errors::InvalidArgument(
+            "indices", SliceDebugString(indices.shape(), bad_i), " = ",
+            indices_vec(bad_i), " is not in [0, ", var.dim_size(0), ")"));
```

Comment:
IMO making this non local is harder to understand.  Can you have `operatior()` return a `Status` (`OK` or `InvalidArgument`) and use `OP_REQUIRES_OK` here?

Reply:
Done